### ISSUE TITLE
support TypedDict (and Required, NotRequired, ReadOnly)

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#154]
- [x] Introduce a new `Type` variant, basic support for class-based `TypedDict` syntax
  - [x] Parse a `TypedDict` specification from the class body
  - [x] Implement / test assignability to `Mapping[str, object]`
  - [x] Make sure that `TypedDict`-based objects do not behave like instances of that class. For example, attribute access should not work.
  - [x] Implement fallback to `_typeshed._type_checker_internals.TypedDictFallback`
  - [ ] Add `TypedDict` samples to property tests
- [x] Type inference for key accesses (`movie['directory'] -> str`)
- [x] Validation of writes to keys (synthesized `__setitem__` method with `key: Literal["key1"], value: ValueType1`?)
- [x] Validation of constructor calls (`Movie(name="Blade Runner", year=1982)`)
- [x] Validation of `{…}` literals to `TypedDict` types
- [x] [Totality](https://typing.python.org/en/latest/spec/typeddict.html#totality)
  - [x] `Required` and `NotRequired`
- [x] Implement / write tests for generic `TypedDict`s
- [x] Implement / write tests for recursive `TypedDict`s
- [x] Support type inference for `instance.get("known_key")`?
- [x] [PEP 705 support](https://peps.python.org/pep-0705/) (`ReadOnly`)
- [ ] #1387 
- [ ] Support `TypedDict`s in all contexts (`return` position, function call arguments, etc) => #168 
- [ ] [PEP 728 support](https://peps.python.org/pep-0728/) ("openness", `extra_items`)
- [ ] Functional syntax (`Movie = TypedDict('Movie', {'name': str, 'year': int})`)
- [ ] Support for `Unpack`
- [ ] Overriding of fields in inheritance
- [ ] Correctness, validation
  - [ ] Implement checks based on [supported and unsupported operations](https://typing.python.org/en/latest/spec/typeddict.html#supported-and-unsupported-operations)
  - [ ] `TypedDict` type objects cannot be used in `isinstance` checks
  - [ ] Only item definitions in the class body (no methods, for example)
  - [ ] Specifying a metaclass is not allowed
  - [ ] Can only subclass `TypedDict`, `TypedDict`- based types and `Generic`
  - [ ] Implement / write tests for subclassing of `TypedDict`-based classes including multiple inheritance
- [ ] Performance
  - [ ] Try out the optimization proposed in https://github.com/astral-sh/ruff/pull/19763#discussion_r2256719857
  - [ ] Add a micro-benchmark for large `TypedDict`s

**Comment by tylerlaprade** [comment]
And `total=False`

**Comment by JelleZijlstra** [comment]
TypedDict has a few interesting properties that may affect core concepts in the type system, so I'll be interested to see how you handle them.

The spec says that a TypedDict must be an instance of `dict` itself, not a subclass. I think that means these test cases should pass:

```python
from typing import Mapping, TypedDict, assert_type

class TD(TypedDict):
    a: int

class SubDict[KT, VT](dict[KT, VT]):
    pass

def f(x: TD | dict[int, str]):
    if isinstance(x, SubDict):
        assert_type(x, SubDict[int, str])
    else:
        assert_type(x, TD | dict[int, str])

def g(x: TD | Mapping[int, str]):
    if isinstance(x, dict):
        assert_type(x, TD | dict[int, str])
    else:
        assert_type(x, Mapping[int, str])
```
(Neither mypy nor pyright currently pass all of this, mostly because they put Any/Unknown in type parameters. I'm not sure there's a good reason for that.)

However, a TypedDict type is not assignable to `dict[T, U]` for any values of T and U, not even to `dict[Any, Any]` (spec: https://typing.python.org/en/latest/spec/typeddict.html#assignability). This will have interesting consequences for the issues in #456: it means that narrowing from `isinstance(x, dict)` can't be implemented by intersecting with some version of `forall K, V. dict[K, V]`.

**Comment by sslivkoff** [comment]
another piece of functionality that would eventually be nice here is the `Unpack` keyword

refs
1. https://typing.python.org/en/latest/spec/callables.html#unpack-kwargs
2. https://peps.python.org/pep-0692/

it's basically a way to use TypedDict's to type the `**kwargs` of a function

this ends up being very useful in practice for balancing python's flexibility with strict type checking

**Comment by carljm** [comment]
Moving this to GA because we feel the current feature-set is adequate for beta (though it would be nice to add a few more things if we have time).

**Comment by AlexWaygood** [comment]
> Moving this to GA because we feel the current feature-set is adequate for beta (though it would be nice to add a few more things if we have time).

(Note that there are many dictionaries that we do not infer as `TypedDict` instances right now despite them being annotated as such, but that's really part of https://github.com/astral-sh/ty/issues/168 rather than this issue -- that issue should definitely be a beta priority!)

**Comment by carljm** [comment]
https://github.com/python/typing/pull/2072 rewrites the TypedDict spec to be much clearer and more concise, which may be useful to anyone working on further features here.

**Comment by JelleZijlstra** [comment]
Specifically a good place to focus on would be the full structure of TypedDict as laid out in the introduction to the proposed spec change (a list of items consisting of key, value type, requiredness, mutability; whether there are extra items; and whether or not the extra items are read-only). The internal representation of TypedDict types likely should follow that structure closely. Then you'll also have to write code to recognize the syntactic forms that create these TypedDicts (e.g., keyword args to the TypedDict constructor; the Required/NotRequired/ReadOnly qualifiers; maybe inheritance, though that can wait). Implementing this structure should provide a good basis for implementing other operations on TypedDicts.

**Comment by tmke8** [comment]
Currently this type-checks on the playground (I don't know whether it also type-checks on main):

```python
from typing import TypedDict

class TD(TypedDict):
    a: str

x: TD = 1
```

Even if full bidirectional type checking isn't implemented yet, wouldn't it be possible to at least check whether the right-hand side is a dictionary?

**Comment by AlexWaygood** [comment]
> Currently this type-checks on the playground (I don't know whether it also type-checks on main):

(The ty playground automatically gets a fresh deploy on each new commit to `main`, so it's usually <10 minutes behind `main`)

**Comment by AlexWaygood** [comment]
anyway: we're working on `TypedDict` features actively, and we will get an improved `TypedDict` assignability implementation delivered to you as soon as we can 😄

**Comment by sharkdp** [comment]
Moving this ticket to GA, but #1387 into the Beta

**Comment by oconnor663** [comment]
Added a checkbox at the top for "Implement `.normalize()` for `TypeDictType`."

**Comment by charliermarsh** [comment]
I'm working on the functional syntax piece.

**Comment by mcrumiller** [comment]
Will this enable narrowing of a `TypedDict` to a `Mapping` when `closed=True` and all of the values are the same type? e.g.:

```python
class A(TypedDict, closed=True):
    a: int
    b: int

def f(d: Mapping[str, int]): ...

my_dict: A = {"a": 1, "b": 2}
f(my_dict)  # Argument to function `f` is incorrect: Expected `Mapping[str, int]`, found `A`
```

**Comment by JelleZijlstra** [comment]
Yes, it should, as specified here: https://typing.python.org/en/latest/spec/typeddict.html#subtyping-with-mapping

**Comment by BenediktMaag** [comment]
Currently i can defined a typed dict, use unpack on a function call but ty will not flag this as an error
[ty playground](https://play.ty.dev/3a108d5e-bc49-4902-be0e-6a73915ad056)
pyright on the other hand flags an invalid keyword argument:
[pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjADRQCqKCAhlQNYBQnVANiwGcBUAJIokMAAosQLCMRjEQACiIIylGgEoAXJygGo8iACMlOzChjc%2Bg4QFk4AYX5C9hqKWLAoAfV%2BoEv7KAsS8wPQAVJHsAO4yaAIWTKwcANpiEtKy8oogALpaUAC0AHxQAHJgKMTuHoYAdE3cji52ysZmIAC8AIz0YDAAFkrdAExaQA)

Imo this is the correct behaviour (at least pre 3.15) because [PEP589 (Typed Dicitionaries)](https://peps.python.org/pep-0589/) states:

The use of a key that is not known to exist should be reported as an error, even if this wouldn’t necessarily generate a runtime type error. 
...
Extra keys included in TypedDict object construction should also be caught.

And then [PEP692 (Unpack for keyword arguments)](https://peps.python.org/pep-0692/) states:

When Unpack is used, type checkers treat kwargs inside the function body as a TypedDict

Is this something that is just not supported yet or is this a design you intentionally use? If so, what is the rational behind it?
I know this might be changing with python 3.15 implementing [PEP728](https://peps.python.org/pep-0728/)

**Comment by sharkdp** [comment]
@BenediktMaag The behavior here is intentional. See https://github.com/python/typing/pull/2272 for the rationale.

We do not plan to change behavior, but in 3.15, you'll be able to use `closed=True` on your TypedDict, which already does what you want in ty today: https://play.ty.dev/437ea9d0-4e99-4ad2-9c46-6f7ac1408384

**Comment by JelleZijlstra** [comment]
> in 3.15, you'll be able to use closed=True on your TypedDict

Or earlier, if you use `typing_extensions.TypedDict`.

**Comment by BenediktMaag** [comment]
Thanks for the clarification, i guess ill postpone it until python 3.15

is there a reason you are reordering the keys in the function signature alphabetically instead of keeping the dict order?

<img width="554" height="252" alt="Image" src="https://github.com/user-attachments/assets/3b0c101c-669f-44aa-8ae1-c51a875804ca" />

**Comment by sharkdp** [comment]
The ordering is probably just an (unfortunate) implementation detail. Given that this leads to a worse LSP experience, I think we should probably fix this.
